Added devices API

drbh opened the pull request:
added the devices api to match functionality in [pyzmq](https://learning-0mq-with-pyzmq.readthedocs.io/en/latest/pyzmq/devices/devices.html)

I tried to follow the style of the library - please let me know if there are any changes needed before merging

drbh committed:
add devices apis

drbh committed:
fix typo

drbh committed:
last socket->device typo

tdudz commented:
any updates on this? would like to see it get merged

tdudz commented:
bump on this again, thanks

drbh commented:
@erickt, any update on reviewing/merging this? Thanks

erickt commented on src/lib.rs:
```diff
@@ -95,6 +96,34 @@ impl SocketType {
+            QUEUE => zmq_sys::ZMQ_QUEUE,
+        };
+        raw as c_int
+    }
```
Could this return a result rather than a panic?

Also, could you add a newline before this line?

drbh committed:
address PR comments

updated `from_raw` to return a result rather than a panic, and added a new line between `to_raw` and `from_raw`

erickt reviewed (changes requested):
Looking through the [pyzmq API docs](https://pyzmq.readthedocs.io/en/latest/api/zmq.devices.html), it looks like the device api was dropped in ZMQ 3.2, and replaced with http://api.zeromq.org/master:zmq-proxy. We do already have support for this in https://github.com/erickt/rust-zmq/blob/eede4315e10eccddec4348561c42ded534e6001e/src/lib.rs#L1138.

I wouldn't consider myself the main maintainer of this library (cc @rotty), but it looks like we're currently depending on zeromq 4.1 and above, so could you switch to `proxy`? 

Also, do you know if zmq 4.1 and above actually exposes `zmq_device`? It might not actually be exposed, which might cause a compilation issue.

Finally, if we do decide to support this device API, can you add some tests for it?

tdudz commented:
ah ok i didn't see that device was deprecated. ill try switching over to proxy, thanks

erickt commented:
@tdudz no problem! Please let me know if you still need this functionality, and we can try to come up with a way to implement it (since it seems pyzmq somehow does it)

tdudz commented:
@erickt i just tested my project using proxy and it seems to work fine. sorry for all the trouble, i should have read the docs a little closer! appreciate the help though and thanks for the great zmq package

drbh commented:
closing since original changes targeted deprecated apis and this pr is old/stale. thanks for the original review
